tests/tests/filter.cpp

```diff
@@ -0,0 +1,532 @@
+  }([](int &) -> Error { throw 0; })); // disallow binding lvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t const &, decltype(fn),
+                             decltype(onError)>;
+  }([](int &) -> bool { throw 0; })); // disallow removing const
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t const &, decltype(truePred),
+                             decltype(fn)>;
+  }([](int &) -> Error { throw 0; })); // disallow removing const
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([](int &) -> bool { throw 0; })); // disallow lvalue binding to rvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([](int &) -> Error { throw 0; })); // disallow lvalue binding to rvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([](std::string) -> bool { throw 0; })); // wrong type
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([](std::string) -> Error { throw 0; })); // wrong type
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([]() -> bool { throw 0; })); // wrong arity
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([]() -> Error { throw 0; })); // wrong arity
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([](int, int) -> bool { throw 0; })); // wrong arity
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([](int, int) -> Error { throw 0; })); // wrong arity
+
+  // rvalue operand
+  // --------------
+
+  // TODO: fix these two if needed:
+  // static_assert([](auto &&fn) constexpr {
+  //   return monadic_invocable<filter_t, operand_t &&, decltype(fn),
+  //                            decltype(onError)>;
+  // }([](int const &&) -> bool { throw 0; })); // allow binding const rvalue
+  // static_assert([](auto &&fn) constexpr {
+  //   return monadic_invocable<filter_t, operand_t &&, decltype(truePred),
+  //                            decltype(fn)>;
+  // }([](int const &&) -> Error { throw 0; })); // allow binding const rvalue
+
+  // TODO: add rvalue stuff here :-O
```

Comment:
I fear this will grow out of control very quickly. On the other hand we do want to check as many use-cases as possible.

Reply:
With the macros above, it's not that scary anymore imo.